File: scripts/serial_read.rb
# data collection from serial port
# 2019-11-14 Naoki F., AIT
# New BSD License is applied. See COPYING file for details.

require 'rubygems'
require 'serialport'
PORT = "COM5" # modify the port according to environment
BLOCK_SIZE = 4096

# Check Arguments
file_name = ARGV[0]
data_size = ARGV[1] ? ARGV[1].to_i : 0
prog_size = data_size / 50
if ! file_name || data_size <= 0
  STDERR.puts "Usage: ruby serial_read.rb FILE_NAME data_size"
  exit 1
end

# Open Serial Port and Output File
begin
  sp = SerialPort.new(PORT, 3000000, 8, 1, 0)
  sp.read_timeout = 1000
rescue => e
  STDERR.puts "!! failed to open serial port (%s). stop." % e.message
  exit 1
end
begin
  out = open(file_name, "wb:ASCII-8BIT")
rescue => e
  STDERR.puts "!! failed to open output file (%s). stop." % e.message
  exit 1
end

# Read 
total = 0
start = nil
while total < data_size
  data = sp.read(BLOCK_SIZE)
  next if ! data
  out.write(data)
  if total / prog_size != (total + data.size) / prog_size
    print '#'
    STDOUT.flush
  end
  total += data.size
  start = Time.now if !start && total >= data_size / 2
end
elapsed = Time.now - start
sp.close
out.close
puts
puts "Elapsed Time(s): %.3f" % (elapsed * 2)
puts "Transfer Rate (kbit/s): %.3f" % (total / elapsed / 250)

File: scripts/serial_read.py
# data collection from serial port, Python alternative
# 2019-11-14 -> 2020-11-13 Naoki F., AIT
# New BSD License is applied. See COPYING file for details.

import sys
import time
import serial
PORT = 'COM5' # modify the port according to environment
BLOCK_SIZE = 4096

# Check Arguments
if len(sys.argv) != 3:
    print('Usage: python serial_read.py FILE_NAME data_size', file=sys.stderr)
    sys.exit(1)

file_name = sys.argv[1]
data_size = int(sys.argv[2])
prog_size = data_size // 50

# Open Serial Port and Output File
try:
    ser = serial.Serial(PORT, 3000000, timeout=1)
except Exception as e:
    print(f'!! failed to open serial port ({e}). stop.', file=sys.stderr)
    sys.exit(1)

try:
    out = open(file_name, mode='wb')
except Exception as e:
    print(f'!! failed to open output file ({e}). stop.', file=sys.stderr)
    sys.exit(1)

# Read
total = 0
start = None
while total < data_size:
    data = ser.read(BLOCK_SIZE)
    if len(data) == 0:
        continue
    out.write(data)
    if total // prog_size != (total + len(data)) // prog_size:
        print('#', file=sys.stdout, end='')
        sys.stdout.flush()
    total += len(data)
    if start is None and total >= data_size // 2:
        start = time.time()

elapsed = time.time() - start
ser.close()
out.close()
print()
print('Elapsed Time(s): {:.3f}'.format(elapsed * 2))
if elapsed != 0:
    print('Transfer Rate (kbit/s): {:.3f}'.format(total / elapsed / 250))
